windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -0,0 +1,1348 @@
+begin
+  if (FErrorMessage <> '') and FShowErrors then
+    KL.Log(FErrorMessage); // TODO: #10210 Log to Sentry
+
+  if FParams.Result = oucShutDown then
+    ShutDown;
+
+  FIdle.Free;
+  FUpdateAvailable.Free;
+  FDownloading.Free;
+  FWaitingRestart.Free;
+  FInstalling.Free;
+  FRetry.Free;
+  FPostInstall.Free;
+
+  // TODO: #10210 remove debugging comments
+  //KL.Log('TUpdateStateMachine.Destroy: FErrorMessage = '+FErrorMessage);
+  //KL.Log('TUpdateStateMachine.Destroy: FParams.Result = '+IntToStr(Ord(FParams.Result)));
+
+  inherited Destroy;
+end;
+
+
+procedure TUpdateStateMachine.SavePackageUpgradesToDownloadTempPath;
+var
+  i: Integer;
+  StringList : TStringList;
+begin
+  StringList := TStringList.Create;
+  try
+    for i := 0 to High(FParams.Packages) do
+      if FParams.Packages[i].NewID <> '' then
+        StringList.Add(FParams.Packages[i].NewID+'='+FParams.Packages[i].ID);
+    if StringList.Count > 0 then
+      StringList.SaveToFile(DownloadTempPath + SPackageUpgradeFileName);
+  finally
+    StringList.Free;
+  end;
+end;
+
+procedure TUpdateStateMachine.ShutDown;
+begin
+  if Assigned(Application) then
+    Application.Terminate;
+end;
+
+{ TOnlineUpdateSharedData }
+
+constructor TOnlineUpdateSharedData.Create(AParams: TUpdateStateMachineParams);
+begin
+  inherited Create;
+  FParams := AParams;
+end;
+
+function TOnlineUpdateSharedData.Params: TUpdateStateMachineParams;
+begin
+  Result := FParams;
+end;
+
+function TUpdateStateMachine.SetRegistryState(Update : TUpdateState): Boolean;
+var
+  UpdateStr : string;
+  Registry: TRegistryErrorControlled;
+begin
+  Result := False;
+  Registry := TRegistryErrorControlled.Create;
+
+  try
+    Registry.RootKey := HKEY_CURRENT_USER;
+
+    if not Registry.OpenKey(SRegKey_KeymanEngine_CU, True) then
+    begin
+      // TODO: #10210 Log to Sentry
+      KL.Log('Failed to open registry key: ' + SRegKey_KeymanEngine_CU);
+      Exit;
+    end;
+
+    try
+      UpdateStr := GetEnumName(TypeInfo(TUpdateState), Ord(Update));
+      Registry.WriteString(SRegValue_Update_State, UpdateStr);
+      Result := True;
+    except
+      on E: Exception do
+      begin
+        // TODO: #10210 Log to Sentry
+        KL.Log('Failed to write to registry: ' + E.Message);
+      end;
+    end;
+
+  finally
+    Registry.Free;
+  end;
+
+end;
+
+function TUpdateStateMachine.CheckRegistryState: TUpdateState;  // I2329
+var
+  UpdateState: TUpdateState;
+  Registry: TRegistryErrorControlled;
+
+begin
+  // We will use a registry flag to maintain the state of the background update
+
+  // check the registry value
+  Registry := TRegistryErrorControlled.Create;  // I2890
+  try
+    Registry.RootKey := HKEY_CURRENT_USER;
+    if Registry.OpenKeyReadOnly(SRegKey_KeymanEngine_CU) and Registry.ValueExists(SRegValue_Update_State) then
+    begin
+      UpdateState := TUpdateState(GetEnumValue(TypeInfo(TUpdateState), Registry.ReadString(SRegValue_Update_State)));
```

Comment:
For a more robust outcome, we should bounds check UpdateState and also wrap the registry calls in a try/except